### PR TITLE
[The Royal We] Allow StakingEscrow initialization from a separate funding source

### PR DESCRIPTION
**Pull request by cygnusv**
A very minor change to allow the DAO agent to activate the contract without giving tokens to it.

### CODE REVIEW

File: tests/contracts/base/test_issuer.py [hunk 1]
```diff
@@ -86,9 +86,16 @@ def calculate_second_phase_reward(locked, total_locked, locked_periods):
...
     with pytest.raises((TransactionFailed, ValueError)):
-        tx = issuer.functions.initialize(0).transact({'from': staker})
+        tx = issuer.functions.initialize(0, staker).transact({'from': staker})
         testerchain.wait_for_receipt(tx)
-    tx = issuer.functions.initialize(economics.erc20_reward_supply).transact({'from': creator})
+
+    # Only owner can initialize
+    with pytest.raises((TransactionFailed, ValueError)):
+        tx = issuer.functions.initialize(economics.erc20_reward_supply, staker).transact({'from': creator})
+        testerchain.wait_for_receipt(tx)
```
Comment (vzotova) [inline tests/contracts/base/test_issuer.py:95]: I guess here should be other comment, something related to that staker has no tokens

File: nucypher/blockchain/eth/agents.py [hunk 2]
```diff
@@ -167,21 +167,24 @@ def get_allowance(self, owner: ChecksumAddress, spender: ChecksumAddress) -> NuN
     @contract_api(TRANSACTION)
     def increase_allowance(self,
                            sender_address: ChecksumAddress,
-                           target_address: ChecksumAddress,
+                           spender_address: ChecksumAddress,
                            increase: NuNits
                            ) -> TxReceipt:
-        contract_function: ContractFunction = self.contract.functions.increaseAllowance(target_address, increase)
-        receipt: TxReceipt = self.blockchain.send_transaction(contract_function=contract_function, sender_address=sender_address)
+        """Increase the allowance from a sender address to a spender address"""
```
Comment (derekpierre) [inline nucypher/blockchain/eth/agents.py:173]:
```suggestion
        """Increase the allowance of a spender address funded by a sender address"""
```

File: nucypher/cli/commands/deploy.py [hunk 3]
```diff
@@ -528,7 +530,8 @@ def allocations(general_config, actor_options, allocation_infile, gas):
 @group_actor_options
 @option_target_address
 @click.option('--value', help="Amount of tokens to transfer in the smallest denomination", type=click.INT)
-def transfer_tokens(general_config, actor_options, target_address, value):
+@click.option('--approve', help="Approve token allowance, instead of transfer", is_flag=True, default=False)
```
Comment (derekpierre) [inline nucypher/cli/commands/deploy.py:533]: 1/4 Perhaps something more descriptive...? Approve by itself feels too general.

```suggestion
@click.option('--allowance-transfer', help="Transfer tokens from target address on behalf of owner", is_flag=True, default=False)
```

File: nucypher/cli/commands/deploy.py [hunk 4]
```diff
@@ -528,7 +530,8 @@ def allocations(general_config, actor_options, allocation_infile, gas):
...
+def transfer_tokens(general_config, actor_options, target_address, value, approve):
```
Comment (derekpierre) [inline nucypher/cli/commands/deploy.py:534]: 2/4

```suggestion
def transfer_tokens(general_config, actor_options, target_address, value, allowance_transfer):
```

File: nucypher/cli/commands/deploy.py [hunk 5]
```diff
@@ -543,11 +546,23 @@ def transfer_tokens(general_config, actor_options, target_address, value):
         stake_value_range = click.FloatRange(min=0, clamp=False)
         value = NU.from_tokens(click.prompt(PROMPT_TOKEN_VALUE, type=stake_value_range))
 
-    confirmation = CONFIRM_TOKEN_TRANSFER.format(value=value,
-                                                 deployer_address=deployer_address,
-                                                 target_address=target_address)
+    value = NuNits(int(value))
+    if approve:
```
Comment (derekpierre) [inline nucypher/cli/commands/deploy.py:550]: 3/4

```suggestion
    if allowance_transfer:
```

File: nucypher/cli/commands/deploy.py [hunk 6]
```diff
@@ -543,11 +546,23 @@ def transfer_tokens(general_config, actor_options, target_address, value):
...
+                                                      deployer_address=deployer_address,
+                                                      spender_address=target_address)
+    else:
+        confirmation = CONFIRM_TOKEN_TRANSFER.format(value=value,
+                                                     deployer_address=deployer_address,
+                                                     target_address=target_address)
     click.confirm(confirmation, abort=True)
-    receipt = token_agent.transfer(amount=int(value), sender_address=deployer_address, target_address=target_address)
+
+    if approve:
```
Comment (derekpierre) [inline nucypher/cli/commands/deploy.py:560]: 4/4

```suggestion
    if allowance_transfer:
```

File: tests/acceptance/cli/test_deploy.py [hunk 7]
```diff
@@ -153,6 +150,25 @@ def test_transfer_tokens(click_runner, registry_filepath):
...
 
+    # Let's approve an allowance to a random spender
+    spender_address = get_random_checksum_address()
+    owner_address = testerchain.client.accounts[0]
+    assert token_agent.get_allowance(spender=spender_address, owner=owner_address) == 0
+
+    command = ['transfer-tokens',
+               '--target-address', spender_address,
+               '--value', 42,
+               '--approve',
```
Comment (derekpierre) [inline tests/acceptance/cli/test_deploy.py:161]: Will need to update, if you accept my suggestion about renaming the argument.

**Comment by mswilkison** [comment]
This PR preserves our ability to activate the contracts independently of the DAO if we decide to go that route, correct?

File: nucypher/cli/commands/deploy.py (hunk 3 above)
Comment (derekpierre) [reply nucypher/cli/commands/deploy.py:533]: ^ A simpler alternative to `--allowance-transfer` (original comment) could be `--allowance` (updated comment)

File: tests/acceptance/cli/test_deploy.py (hunk 7 above)
Comment (cygnusv) [reply tests/acceptance/cli/test_deploy.py:161]:
```suggestion
               '--allowance',
```

File: nucypher/blockchain/eth/agents.py [hunk 8]
```diff
@@ -167,21 +167,24 @@ def get_allowance(self, owner: ChecksumAddress, spender: ChecksumAddress) -> NuN
...
     @contract_api(TRANSACTION)
-    def approve_transfer(self, amount: NuNits,
-                         target_address: ChecksumAddress,
+    def approve_transfer(self,
+                         amount: NuNits,
+                         spender_address: ChecksumAddress,
                          sender_address: ChecksumAddress
                          ) -> TxReceipt:
-        """Approve the transfer of tokens from the sender address to the target address."""
+        """Approve the spender address to transfer an amount of tokens on behalf of the sender address"""
```
Comment (KPrasch) [inline nucypher/blockchain/eth/agents.py:185]: ...eh?  Spender v. Sender is confusing here IMO.

File: tests/acceptance/cli/test_deploy.py [hunk 9]
```diff
@@ -153,6 +150,25 @@ def test_transfer_tokens(click_runner, registry_filepath):
...
+    assert token_agent.get_allowance(spender=spender_address, owner=owner_address) == 0
+
+    command = ['transfer-tokens',
+               '--target-address', spender_address,
+               '--value', 42,
+               '--allowance',
+               '--registry-infile', registry_filepath,
+               '--provider', TEST_PROVIDER_URI]
+
+    user_input = '0\n' + 'Y\n' + 'Y\n'
```
Comment (KPrasch) [inline tests/acceptance/cli/test_deploy.py:165]: Use the `YES` constant.

**Comment by KPrasch** [comment]
:+1: - A few small comments, needs a rebase over main @ #2151. Otherwise, looks good to merge.

File: nucypher/blockchain/eth/agents.py (hunk 8 above)
Comment (cygnusv) [reply nucypher/blockchain/eth/agents.py:185]: I understand the potential confusion, but not sure how to proceed here. "spender" is the term used by the ERC20 implementation we use, by OpenZeppelin. And "sender" is the term we use in all the agents' methods.

File: tests/acceptance/cli/test_deploy.py (hunk 9 above)
Comment (cygnusv) [reply tests/acceptance/cli/test_deploy.py:165]: ✔️

**Pull request merged**